### PR TITLE
[MNT] Move haar wavelet transform to aeon.utils.numba package

### PR DESCRIPTION
**Pull request by CodeLionX**
#### Reference Issues/PRs

None

#### What does this implement/fix? Explain your changes.

In preparation of adding the DWT-MLEAD anomaly detection method, I moved the DFT implementation for `aeon.transformations.collections.DWTTransformer` to a common place in `aeon.utils`.

- Moves the implementation of the Haar wavelet to a separate module `aeon.utils.numba.wavelets`
- Applies `@njit` to the implementation
- Replaces `DWTTransformer`-code with a call to the util-method.

#### Does your contribution introduce a new dependency? If yes, which one?

Nope

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For developers with write access

- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.  
  ⬆️  **should I do this?**

### CODE REVIEW

**Review by MatthewMiddlehurst — APPROVED** [review]
Good work. Fine with this but will leave with others to comment as well before merging.

What is the reasoning for putting this is the numba utils instead of just keeping it as a transformer using the more efficient functions?

**Comment by CodeLionX** [comment]
> Good work. Fine with this but will leave with others to comment as well before merging.
> 
> What is the reasoning for putting this is the numba utils instead of just keeping it as a transformer using the more efficient functions?

I wanted to move it to a common utils place because I also need the Haar wavelet transform for the DWT-MLEAD anomaly detector here: https://github.com/aeon-toolkit/aeon/pull/1580/files#diff-eba996b1b97d4ad19377a87a265148780791ecb1f2940fc38dc28c5307ce836bR168

**Comment by MatthewMiddlehurst** [comment]
Unless the function is also numba compiled it could also use the transformer. Unless it is being called multiple times, I dont think the interface would cause any significant loss in performance.

Not against, just checking on the need for both.

**Comment by CodeLionX** [comment]
Unfortunately, I can't use the transformer directly in DWT-MLEAD. As you can see in this PR, the transformer just collects the `wave_coeffs` while DWT-MLEAD needs the approximate and detail coefficients from almost all the levels.

**Review by baraline — COMMENTED** [review]
I'm not against these changes either, but wouldn't defining these numba functions inside the file of the original transformer and calling these (and not the transformer) for your other needs accomplish the same thing ? Or is it not as “clean” API wise, maybe ?

File: aeon/transformations/collection/dwt.py [hunk 1]
```diff
@@ -52,8 +51,9 @@ def _transform(self, X, y=None):
         Xt : 3D np.ndarray of shape = [n_cases, n_channels, n_timepoints]
             collection of transformed time series
         """
-        n_cases, n_channels, n_timepoints = X.shape
-        _X = np.swapaxes(X, 0, 1)
+        _X = np.array(X, dtype=np.float_)
+        n_cases, n_channels, n_timepoints = _X.shape
+        _X = np.swapaxes(_X, 0, 1)
```
[the remark this answers is not in the document]
Comment (baraline) [reply aeon/transformations/collection/dwt.py:56]: NVM, collections don't (yet?) have axis parameter.

**Comment by CodeLionX** [comment]
> I'm not against these changes either, but wouldn't defining these numba functions inside the file of the original transformer and calling these (and not the transformer) for your other needs accomplish the same thing ? Or is it not as “clean” API wise, maybe ?

Sure, I can move the Haar transform implementation to the transformer file. I did not see this as a nice architectural decision, though. Do you do this kind of calls (one module calling the util-functions of other modules) in other places as well?

Since I am new to the project, I will adapt to your API/architecture style. Just let me know.

**Comment by MatthewMiddlehurst** [comment]
No real solid decision AFAIK. I would say we try to avoid it, but transformations is where the package gets interconnected, as pretty much all other modules interact with it and data transformation in some way. It could go in either currently, IMO.

Unrelated, but both it and `utils` is just kind of a mess of legacy junk currently, me and Tony are clearing it a bit but feel free to give your input on where you think this and other bits should go.

**Comment by CodeLionX** [comment]
I'm fine with both, but would like to have a consistent way of doing this. If there is no single way, we should define rules on where to place the functions in which circumstances, e.g.:

- funcs only used in the transformer ➡️ transformer-file
- funcs that are used in multiple transformers ➡️ `aeon.transformations.common` (or similar)
- funcs that are used in different global modules ➡️ `aeon.utils`
- ...

**Comment by MatthewMiddlehurst** [comment]
@TonyBagnall thoughts? I'm not too fussed.

**Comment by MatthewMiddlehurst** [comment]
Think Tony is busy and I don't really want to hold this up. I'm fine with shared code going into `utils` like this. I would probably do the following:

- funcs only used in the transformer ➡️ transformer-file
- funcs that are used in multiple transformers located in the same file  ➡️ transformer-file
- funcs that are used in multiple transformers in different files  ➡️ aeon.utils
- funcs that are used in different global modules ➡️ aeon.utils

Happy to discuss more or you can just merge. For functions like DFT/FFT which are used in multiple transforms, I'm not sure if some sort of common transformer function package is better or just have it in utils. Probably utils like you have done here.

**Comment by CodeLionX** [comment]
Alright, works for me. We can move common functionality to `aeon.utils` for now, and if we encounter too much stuff to move, we can think about creating a separate common-module in `aeon.transformers`.

Thanks for the discussion 👍🏼

**Review by TonyBagnall — COMMENTED** [review]
LGTM

**Comment by TonyBagnall** [comment]
hi, sorry, catching up on lots here. I am not dogmatic about design tbh. Generally agree with @MatthewMiddlehurst . I think that 1) If you implement an interface/extend a base class, go in that package 
2) go in the estimator file if they are very specific to that estimator and not likely to be useful elsewhere
3) go in utils if they are decoupled from any specific task interface/base class.

**Comment by TonyBagnall** [comment]
(and of course its already merged, lol)

**Pull request merged**